oauthbearer.go

```diff
@@ -61,3 +64,121 @@ func (a *oauthBearerClient) Next(challenge []byte) ([]byte, error) {
 func NewOAuthBearerClient(opt *OAuthBearerOptions) Client {
 	return &oauthBearerClient{*opt}
 }
+
+type OAuthBearerAuthenticator func(opts OAuthBearerOptions) *OAuthBearerError
+
+type oauthBearerServer struct {
+	done         bool
+	fail         error
+	authenticate OAuthBearerAuthenticator
+}
+
+func (a *oauthBearerServer) Next(response []byte) (challenge []byte, done bool, err error) {
+	// Per RFC, we cannot just send an error, we need to return JSON-structured
+	// value as a challenge and then after getting dummy response from the
+	// client stop the exchange.
+	if a.fail != nil {
+		if len(response) != 1 && response[0] != 0x01 {
+			return nil, true, errors.New("unexpected response")
+		}
+		return nil, true, a.fail
+	}
+	fail := func(descr string) ([]byte, bool, error) {
```

Comment:
Can we avoid nesting this function by making it an `oauthBearerServer` method?

Reply:
(Also we could name it `invalidRequest` instead of `fail`)

Reply:
Sure.